### ISSUE TITLE
GitHub private repositories fail due to `unable to find any valid known_hosts file, set SSH_KNOWN_HOSTS env variable`

### ISSUE THREAD

**Issue by usiegj00** [pivotal/kpack#1230]
Kpack 0.11.0 [replaced git2go with go-git](#1159) and doing so GitHub private repositories fail due to this error:
`unable to fetch references for repository: unable to find any valid known_hosts file, set SSH_KNOWN_HOSTS env variable`

The same configuration works with kpack-0.9.4 without issue. This appears to be a regression due git2go -> go-git and may be caused by the handling of known_hosts in the context of the underlying SSH connection to GitHub's git repositories. 

In kpack-0.9.4, private repositories on GitHub required a secret that included GitHub's SSH public key(s). For example KNOWNHOSTS64 below would be `ssh-keyscan github.com` encoded base64. This requirement was pieced together from aged documentation references including #413. It is possible that this issue is from some unrelated misconfiguration we have, so we'd like this to be confirmed prior to giving this issue any more attention!

```
apiVersion: v1
data:
  known_hosts: KNOWNHOSTS64
  ssh-privatekey: PRIVATEKEY64
kind: Secret
metadata:
  annotations:
    kpack.io/git: git@github.com
```  

This has been tested by trying identical Image configurations on kpack-0.9.4 (working) and kpack-0.11.0 (not working).

**Comment by chenbh** [comment]
I am actually not sure why it works in 0.9.4. The only thing we read from the secret is the `ssh-privatekey` field: https://github.com/pivotal/kpack/blob/4ce090eb8a72579bd76ea1604b4e83b89f3ae6d2/pkg/secret/volume_secret_reader.go#L31-L41

I wonder if libgit2 just silently trusts unknown hosts, @usiegj00 can you try using the same image configuration on 0.9.4 but without the `known_hosts` field on the secret?

**Comment by usiegj00** [comment]
Yes--kpack 0.9.4 does not need the known_hosts field. I must have added it in attempting to debug the go-git update.

**Comment by chenbh** [comment]
After talking with @tomkennedy513, since this is technically a regression, we're gonna introduce a `--insecure-ssh-allow-unknown-hosts` (or similarly named) flag which will default to `true` to keep the old behaviour. We'll also introduce a way to provide a `known_hosts` file (probably the same as the way you're doing it) to the build.

Then at some time in the future we can default the flag to false and only allow user supplied `known_hosts`

**Comment by rrileyca** [comment]
Hello,

I had this same issue using version `0.11.0` and `0.11.1`, but I managed a *partial* workaround.

I added the `known_hosts` field to the same `git-ssh-auth` `Secret` that is mounted in the kpack build `Pod` at `/var/build-secrets/git-ssh-auth/`. This means it mounts the `known_hosts` field of the `Secret` as file: `/var/build-secrets/git-ssh-auth/known_hosts` in the container.

In the build pod, by specifying the environment variable `SSH_KNOWN_HOSTS` manually, I am able to specify the value `/var/build-secrets/git-ssh-auth/known_hosts` which contains the contents of:

```shell
ssh-keyscan -H github.com > github_knownhosts.txt 
```

I'm currently looking for a way to always create this ENV variable. Adding it through the `Image` seems to create an environment variable called `PLATFORM_ENV_KNOWN_SSH_HOSTS` which doesn't work/help.
